blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,958 @@
+
+	errBucketTypeNotExist = errors.New("bucket type does not exist")
+
+	// ErrBucketInfoNotExist is the error when bucket does not exist
+	ErrBucketInfoNotExist = errors.New("bucket info does not exist")
+)
+
+func init() {
+	var err error
+	_liquidStakingInterface, err = abi.JSON(strings.NewReader(LiquidStakingContractABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// NewLiquidStakingIndexer creates a new liquid staking indexer
+func NewLiquidStakingIndexer(kvStore db.KVStore, blockInterval time.Duration) LiquidStakingIndexer {
+	return &liquidStakingIndexer{
+		blockInterval: blockInterval,
+		kvstore:       kvStore,
+		cache:         newLiquidStakingCache(),
+	}
+}
+
+// Start starts the indexer
+func (s *liquidStakingIndexer) Start(ctx context.Context) error {
+	if err := s.kvstore.Start(ctx); err != nil {
+		return err
+	}
+	return s.loadCache()
+}
+
+// Stop stops the indexer
+func (s *liquidStakingIndexer) Stop(ctx context.Context) error {
+	if err := s.kvstore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PutBlock puts a block into indexer
+func (s *liquidStakingIndexer) PutBlock(ctx context.Context, blk *block.Block) error {
+	// new dirty cache for this block
+	// it's not necessary to use thread safe cache here, because only one thread will call this function
+	// and no update to cache will happen before dirty merge to clean
+	dirty := newLiquidStakingDirty(s.cache.cache)
+	dirty.putHeight(blk.Height())
+
+	// make action map
+	actionMap := make(map[hash.Hash256]*action.SealedEnvelope)
+	for i := range blk.Actions {
+		h, err := blk.Actions[i].Hash()
+		if err != nil {
+			return err
+		}
+		actionMap[h] = &blk.Actions[i]
+	}
+
+	// handle events of block
+	for _, receipt := range blk.Receipts {
+		if receipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+			continue
+		}
+		act, ok := actionMap[receipt.ActionHash]
+		if !ok {
+			return errors.Errorf("action %x not found", receipt.ActionHash)
+		}
+		for _, log := range receipt.Logs() {
+			if log.Address != LiquidStakingContractAddress {
+				continue
+			}
+			if err := s.handleEvent(ctx, dirty, blk, act, log); err != nil {
+				return err
+			}
+		}
+	}
+
+	// commit dirty cache
+	return s.commit(dirty)
+}
+
+// DeleteTipBlock deletes the tip block from indexer
+func (s *liquidStakingIndexer) DeleteTipBlock(context.Context, *block.Block) error {
+	return errors.New("not implemented")
+}
+
+// Height returns the tip block height
+func (s *liquidStakingIndexer) Height() (uint64, error) {
+	return s.cache.getHeight(), nil
+}
+
+// CandidateVotes returns the candidate votes
+func (s *liquidStakingIndexer) CandidateVotes(candidate address.Address) *big.Int {
+	return s.cache.getCandidateVotes(candidate)
+}
+
+// Buckets returns the buckets
+func (s *liquidStakingIndexer) Buckets() ([]*Bucket, error) {
+	vbs := []*Bucket{}
+	for id, bi := range s.cache.getAllBucketInfo() {
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
```

Comment:
The race condition isn't solved by a threadsafe map. The issue still exists if Putblock() is executed when another thread is running in the loop.

Reply:
pls check whether a mutex is necessary for `liquidStakingIndexer` class

Reply:
Indeed, multiple cache reads before and after the commit still require locking.

Reply:
The improvement looks good for me. But one mutex for this module is enough

Reply:
removed the mutex in cache